<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <link href='https://fonts.googleapis.com/css?family=Chivo:900' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen" />
    <link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen" />
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print" />
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <title>Orthograph by mptrsen</title>
  </head>

  <body>
    <div id="container">
      <div class="inner">

        <header>
          <h1>Orthograph</h1>
          <h2>Orthology prediction using a graph-based, reciprocal approach with profile hidden Markov models</h2>
        </header>

        <section id="downloads" class="clearfix">
          <a href="https://github.com/mptrsen/Orthograph/zipball/master" id="download-zip" class="button"><span>Download .zip</span></a>
          <a href="https://github.com/mptrsen/Orthograph/tarball/master" id="download-tar-gz" class="button"><span>Download .tar.gz</span></a>
          <a href="https://github.com/mptrsen/Orthograph" id="view-on-github" class="button"><span>View on GitHub</span></a>
        </section>

        <hr>

        <section id="main_content">

<h1>
<a name="system-requirements" class="anchor" href="#system-requirements"><span class="octicon octicon-link"></span></a>SYSTEM REQUIREMENTS</h1>

<p>Orthograph requires the following software packages to be installed on the
system. The whole package was tested and works with these versions, and newer
versions should be no problem. Downwards compatibility (i.e., older versions)
has not been tested. Should be fine, too, but be wary with MAFFT, its developers
change or drop features frequently across versions.</p>

<table>
<thead><tr>
<th align="left">Package</th>
<th align="left">Version</th>
<th align="left">Download from</th>
</tr></thead>
<tbody>
<tr>
<td align="left">MAFFT</td>
<td align="left">6.951</td>
<td align="left"><a href="http://mafft.cbrc.jp/alignment/software/">http://mafft.cbrc.jp/alignment/software/</a></td>
</tr>
<tr>
<td align="left">HMMer</td>
<td align="left">3.0</td>
<td align="left"><a href="http://hmmer.janelia.org">http://hmmer.janelia.org</a></td>
</tr>
<tr>
<td align="left">NCBI BLAST+</td>
<td align="left">2.2.28+</td>
<td align="left"><a>ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/</a></td>
</tr>
<tr>
<td align="left">Exonerate</td>
<td align="left">2.2.0</td>
<td align="left"><a href="http://www.ebi.ac.uk/%7Eguy/exonerate/">http://www.ebi.ac.uk/~guy/exonerate/</a></td>
</tr>
</tbody>
</table>

<p></p>

<p>Make sure you get the correct versions for your operating system (most packages
are provided as 32 bit and 64 bit versions).</p>

<p>You have the choice between two different database systems: MySQL and SQLite. The former is a full-blown server-client database management system with superior performance and many features, the latter is a small, file-based database library that is intended to be simple and requires zero configuration. Depending on your choice, you need one of these packages:</p>


<table>
<thead><tr>
<th align="left">Package</th>
<th align="left">Version</th>
<th align="left">Download from</th>
</tr></thead>
<tbody>
<tr>
<td align="left">MySQL</td>
<td align="left">5.1</td>
<td align="left"><a href="http://www.mysql.com">http://www.mysql.com</a></td>
</tr>
<tr>
<td align="left">SQLite</td>
<td align="left">3</td>
<td align="left"><a href="http://www.mysql.com">http://www.sqlite.org/download.html</a></td>
</tr>
</tbody>
</table>

<p></p>

<p>If you use some flavor of Linux, these packages are most likely present in your software repositories. It is recommended to install them using <code>apt-get</code>, <code>yum</code>, or whatever package manager your system uses. Additionally, you need the Perl modules <b>DBI</b> and <b>DBD::mysql</b> or <b>DBD::SQLite</b>, depending on your database choice.</p>

<p>Don't worry about your Perl version if you run a reasonably recent Linux or
UNIX system, such as Mac OS X, but see <a href="http://cpansearch.perl.org/src/JWIED/DBD-mysql-2.1028/INSTALL.html#special%20systems">here</a> and <a href="http://forums.mysql.com/read.php?51,389833,389833">there</a> if you are
trying to install the required packages on a Windows system under Cygwin (or
natively, for that matter). You might run into problems getting the MySQL
driver, HMMer3 and Exonerate to work. Really, don't use Windows. </p>

<p>If you get an error like "Can't locate Some/Module.pm in @INC", you need to
update your Perl installation; see <a href="#required-perl-modules">REQUIRED PERL MODULES</a> for more information.</p>

<h1>
<a name="data-requirements" class="anchor" href="#data-requirements"><span class="octicon octicon-link"></span></a>DATA REQUIREMENTS</h1>

<p>Orthograph is an orthology prediction pipeline that attempts to map transcript
sequences to known orthologous groups. It is not meant to infer orthology de
novo among genomes or transcriptomes, but relies on the orthology information in
pre-defined sets. Because of this requirement, it is essential that your input
data conforms to some data standards as listed below.</p>

<p>a) The clusters of orthologous groups (COGs). It consists of a Fasta file that
contains at least one sequence per selected COG. If you get your COGs from
OrthoDB <a href="recommended">1</a>, this is the case. In an OrthoDB file, each sequence
header contains a COG ID, in the case of OrthoDB version 5, it is something
like:</p>

<pre><code>&gt;APISU:gmsk09 ACYPI007477 EOG500005 ACYPI007477-PA origin:RefSeq
</code></pre>

<p>The 'EOG500005' field is the COG ID. The sequence header must also contain an
unambiguous sequence ID. In the above example, it is the fourth field, i.e.,
'ACYPI007477-PA'. </p>

<p>b) The official gene set (OGS) for each of the taxa that are present in the
ortholog set, i.e., contributed sequences to the COGs. The sequences for a
single taxon must be in Fasta format, and all (!) sequences of this taxon that
occur in the COGs must be in the OGS database as well. Their sequence identifier
must be unambiguously identical to the sequence ID in the Fasta file from
OrthoDB. For example, this would be the related sequence for Acyrthosiphon pisum
from the OGS:</p>

<pre><code>&gt;ACYPI007477-PA origin:RefSeq
</code></pre>

<p>The OGS must be present as amino acid sequences, and may be present as
nucleotide sequences. If the latter is missing, no nucleotide sequences can be
output. The sequence identifiers of corresponding nucleotide and amino acid
sequences must be identical.</p>

<h1>
<a name="data-requirements" class="anchor" href="#data-requirements"><span class="octicon octicon-link"></span></a>QUICKSTART GUIDE</h1>

<p>For quickstart instructions, read the supplied README file. A full manual is being written.</p>

<hr>

<p>Orthograph was developed as part of my Diploma thesis at the ZFMK/zmb, Bonn, Germany 
(c) 2011-2014 Malte Petersen <a href="mailto:mptrsen@uni-bonn.de">mptrsen@uni-bonn.de</a></p>
        </section>

        <footer>
          Orthograph is maintained by <a href="https://github.com/mptrsen">mptrsen</a><br>
          This page was generated by <a href="http://pages.github.com">GitHub Pages</a>. Tactile theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.
        </footer>

        
      </div>
    </div>
  </body>
</html>
