<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <link href='https://fonts.googleapis.com/css?family=Chivo:900' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen" />
    <link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen" />
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print" />
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <title>Orthograph by mptrsen</title>
  </head>

  <body>
    <div id="container">
      <div class="inner">

        <header>
          <h1>Orthograph</h1>
          <h2>Orthology prediction using a graph-based, reciprocal approach with profile hidden Markov models</h2>
        </header>

        <section id="downloads" class="clearfix">
          <a href="https://github.com/mptrsen/Orthograph/zipball/master" id="download-zip" class="button"><span>Download .zip</span></a>
          <a href="https://github.com/mptrsen/Orthograph/tarball/master" id="download-tar-gz" class="button"><span>Download .tar.gz</span></a>
          <a href="https://github.com/mptrsen/Orthograph" id="view-on-github" class="button"><span>View on GitHub</span></a>
        </section>

        <hr>

        <section id="main_content">

<h1>
<a name="system-requirements" class="anchor" href="#system-requirements"><span class="octicon octicon-link"></span></a>SYSTEM REQUIREMENTS</h1>

<p>Orthograph requires the following software packages to be installed on the
system. The whole package was tested and works with these versions, and newer
versions should be no problem. Downwards compatibility (i.e., older versions)
has not been tested. Should be fine, too, but be wary with MAFFT, its developers
change or drop features frequently across versions.</p>

<p>You need a database backend, either SQLite (the default) or MySQL. If you
don't know which one to pick, use SQLite. It's easier to install and use, but
doesn't have the capabilities for a centralized server-client setup. In grid
computing environments, this is an advantage since no network connections are
required, but the database is a flat file on the hard drive.</p>

<table>
<thead><tr>
<th align="left">Package</th>
<th align="left">Version</th>
<th align="left">Download from</th>
</tr></thead>
<tbody>
<tr>
<td align="left">Perl</td>
<td align="left">5.14</td>
<td align="left"><a href="http://www.perl.org">http://www.perl.org</a></td>
</tr>
<tr>
<td align="left">SQLite</td>
<td align="left">3.8.2</td>
<td align="left"><a href="http://sqlite.org/download.html">http://sqlite.org/download.html</a></td>
</tr>
<tr>
<td align="left">MySQL</td>
<td align="left">5.6.17</td>
<td align="left"><a href="http://dev.mysql.com/downloads/mysql/">http://dev.mysql.com/downloads/mysql/</a></td>
</tr>
<tr>
<td align="left">MAFFT</td>
<td align="left">6.951</td>
<td align="left"><a href="http://mafft.cbrc.jp/alignment/software/">http://mafft.cbrc.jp/alignment/software/</a></td>
</tr>
<tr>
<td align="left">HMMer</td>
<td align="left">3.0</td>
<td align="left"><a href="http://hmmer.janelia.org">http://hmmer.janelia.org</a></td>
</tr>
<tr>
<td align="left">NCBI BLAST+</td>
<td align="left">2.2.28+</td>
<td align="left"><a>ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/</a></td>
</tr>
<tr>
<td align="left">Exonerate</td>
<td align="left">2.2.0</td>
<td align="left"><a href="http://www.ebi.ac.uk/%7Eguy/exonerate/">http://www.ebi.ac.uk/~guy/exonerate/</a></td>
</tr>
</tbody>
</table>

<p></p>

<p>Make sure you get the correct versions for your operating system (most packages
are provided as 32 bit and 64 bit versions).</p>

<p>If you use some flavor of Linux, these packages are most likely present in your software repositories. It is recommended to install them using <code>apt-get</code>, <code>yum</code>, or whatever package manager your system uses. Additionally, you need the Perl modules <b>DBI</b> and <b>DBD::mysql</b> or <b>DBD::SQLite</b>, depending on your database choice.</p>

<p>Don't worry about your Perl version if you run a reasonably recent Linux or
UNIX system, such as Mac OS X, but see <a href="http://cpansearch.perl.org/src/JWIED/DBD-mysql-2.1028/INSTALL.html#special%20systems">here</a> and <a href="http://forums.mysql.com/read.php?51,389833,389833">there</a> if you are
trying to install the required packages on a Windows system under Cygwin (or
natively, for that matter). You might run into problems getting the MySQL
driver, HMMer3 and Exonerate to work. Really, don't use Windows. </p>

<p>Make sure you have the database interface module DBI for your Perl version as
well as the proper database driver for your database engine of choice
(DBD::mysql or DBD::SQLite)</p>

<p>If you get an error like "Can't locate Some/Module.pm in @INC", you need to
update your Perl installation; see <a href="#required-perl-modules">REQUIRED PERL MODULES</a> for more information.</p>

<h1>
<a name="data-requirements" class="anchor" href="#data-requirements"><span class="octicon octicon-link"></span></a>DATA REQUIREMENTS</h1>

<p>Orthograph is an orthology prediction pipeline that attempts to map transcript
sequences to known orthologous groups. It is not meant to infer orthology de
novo among genomes or transcriptomes, but relies on the orthology information in
pre-defined sets. Because of this requirement, it is essential that your input
data conforms to some data standards as listed below.</p>

<p>a) The clusters of orthologous groups (COGs). It consists of a tab-delimited
file that contains at least one line per COG. If you get your COGs from <a
href"http://orthodb.org/">OrthoDB</a> (recommended), this is the case. In an
OrthoDB file, each line has a number of tab-delimited fields like this:</p>

<pre><code>EOG7M10DZ       AECH19093       AECH19093-PA    Acromyrmex echinatior   Panamanian leafcutter ant       AECHI   1282                    IPR006121,IPR008250,IPR005834,IPR023214</code></pre>

<p>The 'EOG7M10DZ' field is the COG ID. The line must also contain an unambiguous
sequence ID. For OrthoDB version 7, it is always in the third field, i.e.,
'AECH19093-PA'. The sixth field is the taxon shorthand (I know this is odd, but
it's the way OrthoDB 7 outputs. In the future you'll be able to specify the
field).</p>

<p>If you want to create these files yourself, make sure that the COG ID is in
the first tab-separated field, the sequence ID in the third field, and the taxon
shorthand in the sixth field. Any fields in between are irrelevant to Orthograph
but must not be omitted since it relies on their order.</p>

<p>b) The official gene set (OGS) for each of the taxa that are present in the
ortholog set, i.e., contributed sequences to the COGs. The sequences for a
single taxon must be in Fasta format, and all (!) sequences of this taxon that
occur in the COGs must be in the OGS database as well. Their sequence identifier
must be unambiguously identical to the sequence ID in the file from OrthoDB. For
example, this would be the related sequence for Acromyrmex echinatior from the
OGS:</p>

<pre><code>&gt;AECH19093-PA</code></pre>

<p>The OGS must be present as amino acid sequences, and may be present as
nucleotide sequences. If the latter is missing, no nucleotide sequences can be
output for the reference taxa. The sequence identifiers of corresponding
nucleotide and amino acid sequences must be identical.</p>

<p>The helper script <pre>make-ogs-corresponding.pl</pre> is provided that uses
Exonerate to generate 100% corresponding amino acid and nucleotide sequences
for the OGS. It only requires that the identifiers are unambiguously identical.
You can call it like this:</p>

<pre><code>perl make-ogs-corresponding.pl [OPTIONS] PEPTIDEFILE CDSFILE</code></pre>

<p>Provide the OGS on amino acid level as PEPTIDEFILE and on nucleotide level as
CDSFILE. A list of options is printed when you call the script without arguments
or with the <pre>-h</pre> option.</p>

<p>The OGS must be present as amino acid sequences, and may be present as
nucleotide sequences. If the latter is missing, no nucleotide sequences can be
output. The sequence identifiers of corresponding nucleotide and amino acid
sequences must be identical.</p>

<h1>
<a name="data-requirements" class="anchor" href="#data-requirements"><span class="octicon octicon-link"></span></a>QUICKSTART GUIDE</h1>

<p>For quickstart instructions, read the supplied README file. A full manual is being written.</p>

<hr>

<p>Orthograph was developed as part of my Diploma thesis at the ZFMK/zmb, Bonn, Germany 
(c) 2011-2014 Malte Petersen <a href="mailto:mptrsen@uni-bonn.de">mptrsen@uni-bonn.de</a></p>
        </section>

        <footer>
          Orthograph is maintained by <a href="https://github.com/mptrsen">mptrsen</a><br>
          This page was generated by <a href="http://pages.github.com">GitHub Pages</a>. Tactile theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.
        </footer>

        
      </div>
    </div>
  </body>
</html>
